File: ob-zt/ac4a/java/Ac4aInterceptor.java
package hipstershop;

import io.grpc.Metadata;
import io.grpc.ServerCall;
import io.grpc.ServerCallHandler;
import io.grpc.ServerInterceptor;
import io.grpc.Status;
import io.jsonwebtoken.Jwts;
import java.nio.charset.StandardCharsets;

public class Ac4aInterceptor implements ServerInterceptor {

    private static final boolean ENABLED =
        "true".equalsIgnoreCase(System.getenv("ZT_AC4A"));
    private static final String JWT_SECRET =
        getEnv("JWT_SECRET", "dev-secret-change-me");
    private static final Metadata.Key<String> AUTH_KEY =
        Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER);

    @Override
    public <Q, R> ServerCall.Listener<Q> interceptCall(
            ServerCall<Q, R> call, Metadata headers, ServerCallHandler<Q, R> next) {
        if (!ENABLED) return next.startCall(call, headers);

        String auth = headers.get(AUTH_KEY);
        if (auth == null || !auth.startsWith("Bearer ")) {
            call.close(
                Status.UNAUTHENTICATED.withDescription("missing bearer token"),
                new Metadata());
            return new ServerCall.Listener<Q>() {};
        }
        try {
            Jwts.parser()
                .setSigningKey(JWT_SECRET.getBytes(StandardCharsets.UTF_8))
                .parseClaimsJws(auth.substring(7));
        } catch (Exception e) {
            call.close(
                Status.UNAUTHENTICATED.withDescription("invalid token: " + e.getMessage()),
                new Metadata());
            return new ServerCall.Listener<Q>() {};
        }
        return next.startCall(call, headers);
    }

    private static String getEnv(String key, String def) {
        String v = System.getenv(key);
        return (v != null && !v.isEmpty()) ? v : def;
    }
}

File: ob-zt/ac4a/csharp/Ac4aInterceptor.cs
using System;
using System.IdentityModel.Tokens.Jwt;
using System.Text;
using System.Threading.Tasks;
using Grpc.Core;
using Grpc.Core.Interceptors;
using Microsoft.IdentityModel.Tokens;

namespace cartservice.infrastructure
{
    public class Ac4aInterceptor : Interceptor
    {
        private static readonly bool Enabled =
            string.Equals(Environment.GetEnvironmentVariable("ZT_AC4A"), "true",
                StringComparison.OrdinalIgnoreCase);
        private static readonly string JwtSecret =
            Environment.GetEnvironmentVariable("JWT_SECRET") ?? "dev-secret-change-me";

        private void Verify(ServerCallContext context)
        {
            if (!Enabled) return;
            var auth = context.RequestHeaders.GetValue("authorization") ?? string.Empty;
            if (!auth.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
                throw new RpcException(new Status(StatusCode.Unauthenticated, "missing bearer token"));
            var token = auth.Substring(7);
            var handler = new JwtSecurityTokenHandler();
            var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(JwtSecret));
            try
            {
                handler.ValidateToken(token, new TokenValidationParameters
                {
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = key,
                    ValidateIssuer = false,
                    ValidateAudience = false,
                    ClockSkew = TimeSpan.FromSeconds(30)
                }, out _);
            }
            catch (Exception e)
            {
                throw new RpcException(new Status(StatusCode.Unauthenticated,
                    $"invalid token: {e.Message}"));
            }
        }

        public override Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
            TRequest request, ServerCallContext context,
            UnaryServerMethod<TRequest, TResponse> continuation)
        {
            Verify(context);
            return continuation(request, context);
        }

        public override Task ServerStreamingServerHandler<TRequest, TResponse>(
            TRequest request, IServerStreamWriter<TResponse> responseStream,
            ServerCallContext context,
            ServerStreamingServerMethod<TRequest, TResponse> continuation)
        {
            Verify(context);
            return continuation(request, responseStream, context);
        }
    }
}
